cprover_bindings/src/goto_program/typ.rs

```diff
@@ -570,6 +620,121 @@ impl Type {
         }
     }
 
+    /// A transparent type wraps a base type inside a struct, and has the same in-memory layout.
+    /// For example,
+    /// ```rust
+    ///     struct TransparentWrapper { int x };
+    /// ```
+    /// We define it recursively as a type which has exactly one field, which it itself either
+    /// a transparent type, or is a scalar type.
+    pub fn is_transparent_type(&self, st: &SymbolTable) -> bool {
+        (self.is_struct_like() || self.is_union_like())
+            && self.unwrap_transparent_type(st).is_some()
+    }
+
+    /// Given a transparent type (see comment on `Type::is_transparent_type()`),
+    /// extract the type it wraps.
+    pub fn unwrap_transparent_type(&self, st: &SymbolTable) -> Option<Type> {
+        // If the type has components, i.e. is either a union or a struct, recurse into them
+        if self.is_struct_like() || self.is_union_like() {
+            let components = st.lookup_components_in_type(self).unwrap();
+            // For now, only handle the case of transparent structs with one field.
+            // TODO https://github.com/model-checking/kani/issues/837
+            if components.len() != 1 {
+                None
+            } else {
+                match &components[0] {
+                    Padding { .. } => None,
+                    Field { typ, .. } => typ.unwrap_transparent_type(st),
+                }
+            }
+        } else if self.is_scalar() {
+            Some(self.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Get the non-zero-sized fields (including padding) in self
+    pub fn get_non_empty_components<'a>(
+        &self,
+        st: &'a SymbolTable,
+    ) -> Option<Vec<&'a DatatypeComponent>> {
+        if let Some(components) = st.lookup_components_in_type(self) {
+            Some(components.iter().filter(|x| x.sizeof_in_bits(st) != 0).collect())
+        } else {
+            None
+        }
+    }
+
+    /// Two types are structurally equivilent if one can be cast into the other without changing
+    /// any bytes.  For e.g.,
+    /// ```
+    /// struct foo {
+    ///     char x;
+    ///     int y;
+    /// }
+    /// ```
+    /// is structurally equivalent to
+    /// ```
+    /// struct i {
+    ///     int z;
+    /// }
+    ///
+    /// struct bar {
+    ///     char a;
+    ///     struct i b;
+    /// }
+    /// ```
+    /// But, `struct foo` is not structurally equivilent to:
+    /// ```
+    /// __attribute__((packed))
+    /// struct baz {}
+    ///     char x;
+    ///     int y;
+    /// }
+    /// ```
+    /// Since they have different padding.
+    pub fn is_structurally_equivalent_to(&self, other: &Type, st: &SymbolTable) -> bool {
```

Comment:
IMO, break the branches of this function into 2 helper functions: scalar,other and other,other. Eliminates code duplication, shortens the functions.